pingcap/tidb
# executor: kill tableReader for each connection correctly

### What problem does this PR solve?

Issue Number: close a part of #18057 

Problem Summary:
The tableReader in the coprocessor may retry many times to read regions, druing which, it cannot be aware of the cancelation from client. It may take much of memory for a long time. 

### What is changed and how it works?

Proposal: detect `Killed` to stop the goroutine.

What's Changed: should not use `CompareAndSwap` to change the `Killed` before Close(), because many runninig goroutines should detect the `Killed`.

How it Works: the TableReader the detect the `killed` in every loop.
The query may not exit quickly, because the query do some close work.

### Check List 

Tests 

- Integration test 
- Manual test (add detailed scripts or steps below)
test on mocktikv:
```
step1. explain analyze select * from lineitem;
step2. kill tidb sessionID on another client;
to see the connection is cancelled.
```
similarly test on TiFlash is ok.

### Release note 

- To speed up canceling a query
